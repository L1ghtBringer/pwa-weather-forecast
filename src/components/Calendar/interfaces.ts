export interface IBlockDay {
  beforeDay: Date
  afterDay: Date
}

export interface ILocale {
  monthNames: string[]
  weekDayNames: string[]
  startWeek: 'Su' | 'Mo'
  today: string
}

export interface ICalendarProps {
  locale?: ILocale
  years?: number[]
  blockDay?: IBlockDay
  onChange?: (date: Date) => void
  selectedDate?: Date | null
  style?: string
}
