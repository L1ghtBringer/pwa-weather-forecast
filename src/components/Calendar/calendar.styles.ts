import styled from 'styled-components'

export const CalendarCard = styled.div<{ styleUser?: string }>`
  width: 252px;
  font-family: Ubuntu;
  user-select: none;
  background-color: #fff;
  border: 2px solid rgba(128, 131, 164, 0.06);
  box-shadow: 0px 2px 4px rgba(4, 5, 73, 0.16),
    14px 14px 20px rgba(5, 6, 114, 0.08);
  border-radius: 8px;
  ${(props) => props.styleUser};
`

interface IButton {
  padding?: string
  isDate?: boolean
  today?: boolean
  selected?: boolean
  thisMonth?: boolean
}

export const Button = styled.button<IButton>`
  background-color: transparent;
  outline: none;
  border: 0;
  cursor: pointer;
  padding: ${(props) => props.padding};
  ${(props) =>
    props.isDate &&
    `
    width: 33px;
    height: 33px;
    &:hover {
      background-color: rgba(0, 0, 0, 0.08);
      transition: background 0.4s;
    }
  `}
  &:disabled {
    cursor: not-allowed;
    color: rgba(0, 0, 0, 0.25);
    color: ${(props) => props.today && '#373AF550'};
  }
  ${(props) =>
    props.isDate &&
    !props.thisMonth &&
    `
    color: rgba(0, 0, 0, 0.25);
  `}
  ${(props) =>
    props.isDate &&
    props.today &&
    `
    color: #373AF5;
    font-weight: 700;
  `}
  ${(props) =>
    props.isDate &&
    props.selected &&
    `
    color: #fff;
    font-weight: 700;
    border-radius: 50%;
    background-color: #373AF5;
    &:hover{
      background-color: #373AF5;
    }
  `}
`

export const Header = styled.header`
  display: flex;
  padding: 0 1.2em;
  margin: 0 auto;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid #f0f0f0;
  font-size: 14px;
  span {
    line-height: 40px;
    font-weight: 500;
  }
`

export const Table = styled.table`
  margin: 0 auto;
  width: 90%;
  font-size: 14px;
  border-collapse: collapse;
  th {
    font-weight: 600;
    line-height: 40px;
  }
`

export const Footer = styled.footer`
  display: flex;
  justify-content: center;
  border-top: 1px solid #f0f0f0;
  button {
    line-height: 38px;
    color: #373af5;
    font-weight: 600;
  }
  button:disabled {
    color: #373af560;
  }
`
