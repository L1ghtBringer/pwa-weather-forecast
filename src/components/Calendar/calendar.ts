const DAYS_IN_WEEK = 7

const MAX_WEEK_IN_MONTH = 6

const DAYS_IN_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

const WEEK_DAYS_FROM_MONDAY = [6, 0, 1, 2, 3, 4, 5]

const WEEK_DAYS_FROM_MONDAY_EU = [0, 1, 2, 3, 4, 5, 6]

const Month = {
  January: 0,
  February: 1,
  March: 2,
  April: 3,
  May: 4,
  June: 5,
  July: 6,
  August: 7,
  September: 8,
  October: 9,
  Novermber: 10,
  December: 11
}

export const areEqual = (a: Date | undefined, b: Date | undefined): boolean => {
  if (!a || !b) return false

  return (
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate()
  )
}

export const qt = (a: Date | undefined, b: Date | undefined): boolean => {
  if (!a || !b) return false

  return (
    new Date(a.getFullYear(), a.getMonth(), a.getDate()) >
    new Date(b.getFullYear(), b.getMonth(), b.getDate())
  )
}

export const isLeapYear = (year: number): boolean => {
  return !(year % 4 || (!(year % 100) && year % 400))
}

export const getDaysInMonth = (date: Date): number => {
  const month = date.getMonth()
  const year = date.getFullYear()
  const daysInMonth = DAYS_IN_MONTH[month]

  if (isLeapYear(year) && month === Month.February) {
    return daysInMonth + 1
  } else {
    return daysInMonth
  }
}

export const getDayOfWeek = (date: Date, startWeek: 'Su' | 'Mo' = 'Su'): number => {
  const dayOfWeek = date.getDay()

  return startWeek === 'Su'
    ? WEEK_DAYS_FROM_MONDAY_EU[dayOfWeek]
    : WEEK_DAYS_FROM_MONDAY[dayOfWeek]
}

export const getMonthData = (year: number, month: number, startWeek: 'Su' | 'Mo' = 'Su'): Array<Array<(Date| undefined)>> => {
  const result: Array<Array<(Date| undefined)>> = []
  const date: Date = new Date(year, month)
  const daysInMonth: number = getDaysInMonth(date)
  const monthStartsOn: number = getDayOfWeek(date, startWeek)
  let day: number = 1
  let dayAfterMonth: number = 1

  for (let i = 0; i < MAX_WEEK_IN_MONTH; i++) {
    result[i] = []
    for (let j = 0; j < DAYS_IN_WEEK; j++) {
      if (i === 0 && j < monthStartsOn) {
        result[i][j] = new Date(year, month, j - monthStartsOn + 1)
      } else if (day > daysInMonth) {
        result[i][j] = new Date(year, month + 1, dayAfterMonth++)
      } else {
        result[i][j] = new Date(year, month, day++)
      }
    }
  }

  return result
}
