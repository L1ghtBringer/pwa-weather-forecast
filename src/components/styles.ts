import styled from 'styled-components'

export const Input = styled.input`
  width: 100%;
  font-family: ${props => props.theme.font};
  font-weight: normal;
  font-size: 16px;
  line-height: 24px;
  padding: 10px 16px;
  background: rgba(128, 131, 164, 0.06);
  border: 2px solid rgba(128, 131, 164, 0.2);
  border-radius: 8px;
  outline: none;
  color: #2C2D76;
  :hover{
    border: 2px solid #8083A4;
  }
  :focus{
    border: 2px solid #373AF5;
  }
  -webkit-transform: translate3d(0,0,0);
  transition: border 0.4s ease-out;
  cursor: pointer;
`

export const Container = styled.div`
  max-width: 252px;
  width: 100%;
  position: relative;
  img {
    position: absolute;
    pointer-events: none;
  }
  @media ${props => props.theme.media.phone} {
    max-width: 100%;
  }
`
