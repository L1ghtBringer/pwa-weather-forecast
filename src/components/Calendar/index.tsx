import React, { useState } from 'react'
import { ICalendarProps } from './interfaces'
import { Button, CalendarCard, Footer, Header, Table } from './calendar.styles'
import { getMonthData, areEqual, qt } from './calendar'
import { locale } from './locale/en_GB'

const defaultProps: ICalendarProps = {
  locale,
  years: [2020, 2021, 2022, 2023, 2024, 2025],
  onChange: (date: Date) => {}
}

interface IState {
  date: Date
  currentDate: Date
  selectedDate: Date | null
}

const Calendar: React.FC<ICalendarProps> = ({
  years,
  locale,
  selectedDate,
  blockDay,
  onChange,
  style
}) => {
  const [state, setState] = useState<IState>({
    date: selectedDate ?? new Date(),
    currentDate: new Date(),
    selectedDate: selectedDate ?? null
  })

  const year: number = state.date.getFullYear()
  const month: number = state.date.getMonth()

  const monthDate = getMonthData(year, month, locale ? locale.startWeek : 'Su')

  const handlePrevMonthButtonClick = (): void => {
    setState((prevState) => ({
      ...prevState,
      date: new Date(year, month - 1)
    }))
  }

  const handleNextMonthButtonClick = (): void => {
    setState((prevState) => ({
      ...prevState,
      date: new Date(year, month + 1)
    }))
  }

  const handleDayClick = (date: Date): void => {
    setState((prevState) => ({ ...prevState, date, selectedDate: date }))
    onChange!(date)
  }

  const setTodayDay = (): void => {
    if (!disabledDay(new Date()) && onChange) {
      const today = new Date()
      setState((prevState) => ({
        ...prevState,
        date: today,
        selectedDate: today
      }))
      onChange(new Date(today.getFullYear(), today.getMonth(), today.getDate()))
    }
  }

  const disabledDay = (date: Date): boolean => {
    if (!blockDay) return false
    if (qt(date, blockDay.afterDay) || !qt(date, blockDay.beforeDay)) {
      return true
    }
    return false
  }

  return (
    <CalendarCard styleUser={String(style)}>
      <Header>
        <Button
          padding={'.5em'}
          onClick={handlePrevMonthButtonClick}
          disabled={new Date(year, month - 1).getFullYear() < years![0]}
        >
          {'<'}
        </Button>
        <span>{locale!.monthNames[month]}</span>
        <span>{year}</span>
        <Button
          padding={'.5em'}
          onClick={handleNextMonthButtonClick}
          disabled={
            new Date(year, month + 1).getFullYear() > years![years!.length - 1]
          }
        >
          {'>'}
        </Button>
      </Header>
      <Table>
        <thead>
          <tr>
            {locale!.weekDayNames.map((name) => (
              <th key={name}>{name}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {monthDate.map((week, index) => (
            <tr key={index}>
              {week.map((date, index) => (
                <td key={index}>
                  <Button
                    isDate={true}
                    today={areEqual(date, state.currentDate)}
                    selected={areEqual(date, state.selectedDate!)}
                    onClick={() => handleDayClick(date!)}
                    thisMonth={date?.getMonth() === state.date.getMonth()}
                    disabled={disabledDay(date!)}
                  >
                    {date!.getDate()}
                  </Button>
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </Table>
      <Footer>
        <Button disabled={disabledDay(new Date())} onClick={setTodayDay}>
          {locale!.today}
        </Button>
      </Footer>
    </CalendarCard>
  )
}

Calendar.defaultProps = defaultProps

export default Calendar
