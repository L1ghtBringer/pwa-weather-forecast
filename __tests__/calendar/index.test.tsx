import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Calendar from '../../src/components/Calendar'
import { Header, Button } from '../../src/components/Calendar/calendar.styles'
import MockDate from 'mockdate'

configure({ adapter: new Adapter() })

describe('Calendar component test', () => {
  it('Should match the snapshot', () => {
    const calendar = shallow(<Calendar />)
    expect(calendar.html()).toMatchSnapshot()
    expect(calendar.exists()).toBe(true)
  })

  it('Test default date', () => {
    MockDate.set('2021-05-01')
    const calendar = shallow(<Calendar />)
    expect(calendar.find(Header).at(0).text()).toBe('<May2021>')
    MockDate.reset()
  })

  it('Test props date', () => {
    const date = new Date(2021, 4, 1)
    const calendar = shallow(<Calendar selectedDate={date}/>)
    expect(calendar.find(Header).at(0).text()).toBe('<May2021>')
    MockDate.reset()
  })

  it('Prev month', () => {
    MockDate.set('2021-05-01')
    const calendar = shallow(<Calendar />)
    expect(calendar.find(Header).at(0).text()).toBe('<May2021>')
    calendar.find(Button).at(0).simulate('click')
    expect(calendar.find(Header).at(0).text()).toBe('<April2021>')
  })

  it('Prev month disabled', () => {
    MockDate.set('2021-01-01')
    const calendar = shallow(<Calendar years={[2021]} />)
    expect(calendar.find(Header).at(0).text()).toBe('<January2021>')
    expect(calendar.find(Button).at(0).props().disabled).toBe(true)
  })

  it('Next month', () => {
    MockDate.set('2021-04-01')
    const calendar = shallow(<Calendar />)
    expect(calendar.find(Header).at(0).text()).toBe('<April2021>')
    calendar.find(Button).at(1).simulate('click')
    expect(calendar.find(Header).at(0).text()).toBe('<May2021>')
  })

  it('Next month disabled', () => {
    MockDate.set('2021-12-01')
    const calendar = shallow(<Calendar years={[2021]} />)
    expect(calendar.find(Header).at(0).text()).toBe('<December2021>')
    expect(calendar.find(Button).at(1).props().disabled).toBe(true)
  })

  it('Set today', () => {
    let expectedDate: Date | null = null
    const onChange = jest.fn((date: Date) => (expectedDate = date))
    MockDate.set('2021-5-05')
    const calendar = shallow(<Calendar onChange={onChange} />)
    calendar.find(Button).last().simulate('click')
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(expectedDate).toEqual(new Date(2021, 4, 5))
  })

  it('Set day', () => {
    let expectedDate: Date | null = null
    const onChange = jest.fn((date: Date) => (expectedDate = date))
    MockDate.set('2021-5-05')
    const calendar = shallow(<Calendar onChange={onChange} />)
    const selectDate = calendar.find(Button).at(2).text()
    calendar.find(Button).at(2).simulate('click')
    expect(expectedDate).toEqual(new Date(2021, 3, Number(selectDate)))
  })

  it('This month', () => {
    MockDate.set('2021-5-05')
    const calendar = shallow(<Calendar/>)
    expect(calendar.find(Button).at(2).props().thisMonth).toBe(false)
    expect(calendar.find(Button).at(10).props().thisMonth).toBe(true)
  })

  it('Check disabled day', () => {
    const blockDay = {
      beforeDay: new Date(2021, 4, 1),
      afterDay: new Date(2021, 4, 4),
    }
    MockDate.set('2021-5-01')
    const calendar = shallow(<Calendar blockDay={blockDay} />)
    expect(calendar.find(Button).at(2).props().disabled).toBe(true)
    expect(calendar.find(Button).at(8).props().disabled).toBe(true)
    expect(calendar.find(Button).at(9).props().disabled).toBe(false)
    expect(calendar.find(Button).at(11).props().disabled).toBe(false)
    expect(calendar.find(Button).at(12).props().disabled).toBe(true)
  })

  it('Check disabled day set today', () => {
    const blockDay = {
      beforeDay: new Date(2021, 4, 2),
      afterDay: new Date(2021, 4, 4),
    }
    MockDate.set('2021-5-01')
    const calendar = shallow(<Calendar blockDay={blockDay} />)
    const onChange = jest.fn(() => {})
    calendar.find(Button).last().simulate('click')
    expect(onChange).toHaveBeenCalledTimes(0)
  })
})
