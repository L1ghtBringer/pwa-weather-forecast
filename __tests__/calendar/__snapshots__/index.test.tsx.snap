// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component test Should match the snapshot 1`] = `"<div class=\\"sc-bdnxRM dcdeka\\"><header class=\\"sc-dlnjwi gPwwuj\\"><button class=\\"sc-gtsrHT bEaTPc\\">&lt;</button><span>May</span><span>2021</span><button class=\\"sc-gtsrHT bEaTPc\\">&gt;</button></header><table class=\\"sc-hKFxyN cvoswi\\"><thead><tr><th>Su</th><th>Mo</th><th>Tu</th><th>We</th><th>Th</th><th>Fr</th><th>Sa</th></tr></thead><tbody><tr><td><button class=\\"sc-gtsrHT cukvZJ\\">25</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">26</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">27</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">28</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">29</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">30</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">1</button></td></tr><tr><td><button class=\\"sc-gtsrHT eahuWb\\">2</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">3</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">4</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">5</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">6</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">7</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">8</button></td></tr><tr><td><button class=\\"sc-gtsrHT eahuWb\\">9</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">10</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">11</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">12</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">13</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">14</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">15</button></td></tr><tr><td><button class=\\"sc-gtsrHT eahuWb\\">16</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">17</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">18</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">19</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">20</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">21</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">22</button></td></tr><tr><td><button class=\\"sc-gtsrHT eahuWb\\">23</button></td><td><button class=\\"sc-gtsrHT iYCQTw\\">24</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">25</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">26</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">27</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">28</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">29</button></td></tr><tr><td><button class=\\"sc-gtsrHT eahuWb\\">30</button></td><td><button class=\\"sc-gtsrHT eahuWb\\">31</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">1</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">2</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">3</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">4</button></td><td><button class=\\"sc-gtsrHT cukvZJ\\">5</button></td></tr></tbody></table><footer class=\\"sc-eCApnc KehJD\\"><button class=\\"sc-gtsrHT kFkpnb\\">Today</button></footer></div>"`;
